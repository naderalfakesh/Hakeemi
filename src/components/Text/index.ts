import Text from './Text';

export type { Props } from './Text';

export default Text;
