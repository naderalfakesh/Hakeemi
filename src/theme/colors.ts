export const palette = {
  primary: {
    base: 'rgba(0, 188, 212, 1)',
    light: 'rgba(214, 250, 255, 1)',
    dark: 'rgba(144, 180, 206, 1)',
  },
  secondary: {
    base: 'rgba(237, 243, 247, 1)',
  },
  danger: {
    base: 'rgba(255, 87, 34, 1)',
  },
  grey: {
    base: 'rgba(240, 240, 240, 1)',
    light: 'rgba(251, 251, 251, 1)',
    dark: 'rgba(144, 144, 144, 1)',
  },
  white: {
    base: 'rgba(255, 255, 255, 1)',
  },
  black: {
    base: 'rgba(0, 0, 0, 1)',
  },
};

export default {
  primary: palette.primary.base,
  secondary: palette.secondary.base,
  danger: palette.danger.base,
  light: palette.grey.light,
  dark: palette.grey.dark,
  white: palette.white.base,
  black: palette.black.base,
};
