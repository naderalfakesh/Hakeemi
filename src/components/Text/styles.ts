import { StyleSheet, TextStyle } from 'react-native';
import colors, { palette } from '../../theme/colors';

export type Color =
  | 'primary'
  | 'secondary'
  | 'light'
  | 'dark'
  | 'white'
  | 'black';

export type Size = 'button-22' | 'body-16' | 'body-22' | 'head-30';

export default StyleSheet.create({
  base: {
    // fontFamily: 'system',
    includeFontPadding: false,
  },
});

export const sizeTheme: Record<Size, TextStyle> = StyleSheet.create({
  'body-16': {
    fontSize: 16,
    lineHeight: 20,
  },
  'body-22': {
    fontSize: 22,
    lineHeight: 24,
  },
  'button-22': {
    fontSize: 22,
    lineHeight: 28,
  },
  'head-30': {
    fontSize: 30,
    lineHeight: 38,
  },
});

export const colorTheme: Record<Color, TextStyle> = StyleSheet.create({
  primary: {
    color: colors.primary,
  },
  secondary: {
    color: palette.primary.dark,
  },
  light: {
    color: colors.light,
  },
  dark: {
    color: colors.dark,
  },
  white: {
    color: colors.white,
  },
  black: {
    color: colors.black,
  },
});
