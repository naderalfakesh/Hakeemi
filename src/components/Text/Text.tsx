import React, { FC } from 'react';
import {
  TextStyle,
  Text as TextBase,
  StyleSheet,
  TextProps,
} from 'react-native';
import styles, { Color, colorTheme, Size, sizeTheme } from './styles';

export interface Props extends TextProps {
  style?: TextStyle;
  size?: Size;
  color?: Color;
  align?: TextStyle['textAlign'];
  level?: TextStyle['fontWeight'];
  onPress?: () => void;
}

const Text: FC<Props> = ({
  children,
  style,
  size = 'body-16',
  color = 'primary',
  align,
  level = '400',
  onPress,
  ...otherProps
}) => {
  const flatStyle = StyleSheet.flatten([
    styles.base,
    colorTheme[color],
    sizeTheme[size],
    { textAlign: align, fontWeight: level },
    style,
  ]);

  return (
    <TextBase testID="Text" style={flatStyle} onPress={onPress} {...otherProps}>
      {children}
    </TextBase>
  );
};

export default Text;
