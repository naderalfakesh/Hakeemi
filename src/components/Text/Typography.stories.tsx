import React, { Fragment } from 'react';
import { View, StyleSheet } from 'react-native';
import { storiesOf } from '@storybook/react-native';
import Text from './Text';
import { Color } from './styles';

const Container = (storyFn: any) => (
  <View style={styles.container}>{storyFn()}</View>
);

const TextDemo = ({ color }: { color: Color }) => (
  <Fragment>
    <Text color={color} size="body-10">
      Lorem ipsum dolor sit.
    </Text>
    <Text color={color} size="body-14">
      Lorem ipsum dolor sit.
    </Text>
    <Text color={color} size="body-16">
      Lorem ipsum dolor sit.
    </Text>
    <Text color={color} size="body-22">
      Lorem ipsum dolor sit.
    </Text>
    <Text color={color} size="body-20">
      Lorem ipsum dolor sit.
    </Text>
    <Text color={color} size="button-22">
      Lorem ipsum dolor sit.
    </Text>
    <Text color={color} size="head-20">
      Lorem ipsum dolor sit.
    </Text>
    <Text color={color} size="head-24">
      Lorem ipsum dolor sit.
    </Text>
    <Text color={color} size="head-30">
      Lorem ipsum dolor sit.
    </Text>
  </Fragment>
);

storiesOf('Typography', module)
  .addDecorator(Container)
  .add('primary', () => <TextDemo color="primary" />)
  .add('secondary', () => <TextDemo color="secondary" />)
  .add('light', () => <TextDemo color="light" />)
  .add('dark', () => <TextDemo color="dark" />)
  .add('white', () => <TextDemo color="white" />)
  .add('black', () => <TextDemo color="black" />);

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#e5e5e5',
    flexGrow: 1,
    padding: 8,
    justifyContent: 'space-evenly',
  },
});
