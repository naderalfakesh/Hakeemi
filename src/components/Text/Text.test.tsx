import React from 'react';
import { render } from '@testing-library/react-native';
import Text from './Text';
import colors, { palette } from '../../theme/colors';

describe('Button component test', () => {
  it('renders correctly with default params', () => {
    const { toJSON } = render(<Text>Hello</Text>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('overrides default style style', () => {
    const { getByText } = render(<Text style={{ color: 'blue' }}>Hello</Text>);
    const txt = getByText('Hello');
    expect(txt.props.style.color).toBe('blue');
  });

  it('sets font weight and text align', () => {
    const { getByText } = render(
      <Text align="center" level="700">
        Hello
      </Text>,
    );
    const txt = getByText('Hello');
    expect(txt.props.style.textAlign).toBe('center');
    expect(txt.props.style.fontWeight).toBe('700');
  });

  it('renders correctly with different sizes', () => {
    const { rerender, getByText } = render(<Text size="body-16">Hello</Text>);
    const txt = getByText('Hello');
    expect(txt.props.style.fontSize).toBe(16);

    rerender(<Text size="body-10">Hello</Text>);
    expect(txt.props.style.fontSize).toBe(10);

    rerender(<Text size="body-14">Hello</Text>);
    expect(txt.props.style.fontSize).toBe(14);

    rerender(<Text size="body-20">Hello</Text>);
    expect(txt.props.style.fontSize).toBe(20);

    rerender(<Text size="body-22">Hello</Text>);
    expect(txt.props.style.fontSize).toBe(22);

    rerender(<Text size="button-22">Hello</Text>);
    expect(txt.props.style.lineHeight).toBe(28);

    rerender(<Text size="head-24">Hello</Text>);
    expect(txt.props.style.fontSize).toBe(24);

    rerender(<Text size="head-20">Hello</Text>);
    expect(txt.props.style.fontSize).toBe(20);

    rerender(<Text size="head-30">Hello</Text>);
    expect(txt.props.style.fontSize).toBe(30);
  });

  it('renders correctly with different colors', () => {
    const { rerender, getByText } = render(<Text color="primary">Hello</Text>);
    const txt = getByText('Hello');
    expect(txt.props.style.color).toBe(colors.primary);

    rerender(<Text color="secondary">Hello</Text>);
    expect(txt.props.style.color).toBe(palette.primary.dark);

    rerender(<Text color="black">Hello</Text>);
    expect(txt.props.style.color).toBe(colors.black);

    rerender(<Text color="white">Hello</Text>);
    expect(txt.props.style.color).toBe(colors.white);

    rerender(<Text color="light">Hello</Text>);
    expect(txt.props.style.color).toBe(colors.light);

    rerender(<Text color="dark">Hello</Text>);
    expect(txt.props.style.color).toBe(colors.dark);
  });
});
