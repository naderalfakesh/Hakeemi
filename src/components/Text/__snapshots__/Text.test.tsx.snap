// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component test renders correctly with default params 1`] = `
<Text
  style={
    Object {
      "color": "rgba(0, 188, 212, 1)",
      "fontFamily": "System",
      "fontSize": 16,
      "fontWeight": "400",
      "includeFontPadding": false,
      "lineHeight": 20,
      "textAlign": undefined,
    }
  }
  testID="Text"
>
  Hello
</Text>
`;
